libs/entitlement/asm-provider/src/main/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumentationServiceImpl.java

```diff
@@ -48,22 +50,108 @@ public MethodVisitor visitMethod(
-
-                var checkerParameterDescriptors = Arrays.stream(checkerMethodArgumentTypes).map(Type::getDescriptor).toList();
-                var checkMethod = new CheckMethod(Type.getInternalName(checkerClass), checkerMethodName, checkerParameterDescriptors);
-
-                methodsToInstrument.put(methodToInstrument, checkMethod);
+                    var checkerParameterDescriptors = Arrays.stream(checkerMethodArgumentTypes).map(Type::getDescriptor).toList();
+                    var checkMethod = new CheckMethod(Type.getInternalName(checkerClass), checkerMethodName, checkerParameterDescriptors);
 
+                    methodsToInstrument.put(methodToInstrument, checkMethod);
+                }
                 return mv;
             }
         };
         reader.accept(visitor, 0);
         return methodsToInstrument;
     }
 
+    @Override
+    public InstrumentationInfo lookupImplementationMethod(Class<?> implementationClass, Method instrumentMethod, Method checkMethod) {
+        checkMethodIsValid(implementationClass, instrumentMethod);
+
+        var instrumentMethodArguments = Type.getArgumentTypes(instrumentMethod);
+        var checkMethodArguments = Type.getArgumentTypes(checkMethod);
+
+        checkArguments(checkMethodArguments, instrumentMethodArguments, Modifier.isStatic(instrumentMethod.getModifiers()));
+
+        return new InstrumentationInfo(
+            new MethodKey(
+                Type.getInternalName(implementationClass),
+                instrumentMethod.getName(),
+                Arrays.stream(instrumentMethodArguments).map(Type::getInternalName).toList()
+            ),
+            new CheckMethod(
+                Type.getInternalName(checkMethod.getDeclaringClass()),
+                checkMethod.getName(),
+                Arrays.stream(checkMethodArguments).map(Type::getDescriptor).toList()
+            )
+        );
+    }
+
+    private static void checkArguments(Type[] checkMethodArguments, Type[] instrumentMethodArguments, boolean isStatic) {
+        var additionalCheckArgs = (isStatic ? 1 : 2);
+        if (checkMethodArguments.length != instrumentMethodArguments.length + additionalCheckArgs) {
+            throw new IllegalArgumentException("The check method argument count is incorrect");
+        }
+
+        if (checkMethodArguments[0] != Type.getType(Class.class)) {
+            throw new IllegalArgumentException("The first argument of a check method must be the caller Class");
+        }
+
+        for (int i = additionalCheckArgs; i < checkMethodArguments.length; ++i) {
+            if (checkMethodArguments[i] != instrumentMethodArguments[i - additionalCheckArgs]) {
+                throw new IllegalArgumentException("Additional arguments of a check method must be the same as the instrumented method");
+            }
+        }
+    }
+
+    private static void checkMethodIsValid(Class<?> implementationClass, Method targetMethod) {
```

Comment:
This is an unfortunate name collision, since this method isn't related to `check$` methods. I can't think of a better name though, unless you want to go for `checkThatMethodIsValid` or similar.

Reply:
Yes, or `ensureMethodIsValid`